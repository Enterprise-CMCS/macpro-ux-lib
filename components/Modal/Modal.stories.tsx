import React, { useState } from "react";
import { Button } from "../Button/Button";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import { Modal } from "./Modal";

export default {
  title: "COMPONENTS/Modal",
  component: Modal,
  argTypes: {
    description: {
      description:
        "The description text that appears on the modal beneath the heading.",
    },
    heading: {
      description: "The heading text that appears on the modal.",
    },
    id: {
      control: false,
      description:
        "The modal ID, used to automatically create related ARIA tags.",
    },
    isOpen: {
      control: false,
      description:
        "A boolean flag that shows/hides the modal, set by the Modal's parent component.",
    },
    modalType: {
      description:
        "The type of Modal. Options include `default`, `large`, or `forcedAction`.",
    },
    onClose: {
      description:
        "The Modal calls this function when the user clicks the close button.",
    },
    primaryButtonText: {
      description: "Text shown on the primary button.",
    },
    primaryOnClick: {
      description:
        "The modal calls this function when the user clicks the primary button.",
    },
    secondaryButtonText: {
      description: "Text shown on the secondary button.",
    },
    secondaryOnClick: {
      description:
        "The modal calls this function when the user clicks the secondary button.",
    },
  },
  args: {
    description: "You have unsaved changes that will be lost.",
    heading: "Are you sure you want to continue?",
    id: "example-modal",
    modalType: "default",
    primaryButtonText: "Primary Button",
    secondaryButtonText: "Secondary Button",
  },
} as ComponentMeta<typeof Modal>;

const Template: ComponentStory<typeof Modal> = ({ ...args }) => {
  const [isOpen, setOpen] = useState(false);
  const toggleModal = () => {
    setOpen(!isOpen);
  };

  return (
    <>
      <Button buttonText="Open modal" onClick={toggleModal} />
      <Modal
        {...args}
        isOpen={isOpen}
        onClose={toggleModal}
        primaryOnClick={toggleModal}
        secondaryOnClick={toggleModal}
      />
    </>
  );
};

export const Default = Template.bind({});
Default.args = {};

export const Large = Template.bind({});
Large.args = { modalType: "large" };

export const ForcedAction = Template.bind({});
ForcedAction.args = { modalType: "forcedAction" };
