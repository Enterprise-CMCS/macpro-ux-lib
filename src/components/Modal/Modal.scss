.usa-modal-overlay {
  z-index: 10;
}