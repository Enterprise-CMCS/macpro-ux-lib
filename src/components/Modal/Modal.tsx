import { Button } from "../Button/Button";
import { Icon } from "../Icon/Icon";
import React from "react";

type ModalType = "default" | "large" | "forcedAction";

interface Props {
  description: string;
  heading: string;
  id?: string;
  isOpen?: boolean;
  modalType?: ModalType;
  onClose?: () => void;
  primaryButtonText: string;
  primaryOnClick: () => void;
  secondaryButtonText: string;
  secondaryOnClick: () => void;
}

/**
 * Modal
 * @callback [onClose]
 * @callback primaryOnClick
 * @callback secondaryOnClick
 * @param {boolean}    [isOpen]             Sets whether or not the modal is visible and open.
 * @param {ModalType}  [modalType]          The type of Modal. Options include "default", "large", or "forcedAction".
 * @param {string}     [id]                 String used for the modal ID and related ARIA tags.
 * @param {string}     description          Description text that appears on the modal beneath the heading.
 * @param {string}     heading              Heading text that appears at the top of the modal.
 * @param {string}     primaryButtonText    Text shown on the primary button.
 * @param {string}     secondaryButtonText  Text shown on the secondary button.
 */

export const Modal: React.FC<Props> = ({
  description,
  heading,
  isOpen = false,
  modalType = "default",
  onClose,
  primaryButtonText,
  primaryOnClick,
  secondaryButtonText,
  secondaryOnClick,
}) => {
  return (
    <>
      {isOpen && (
        <div
          className="usa-modal-wrapper"
          role="dialog"
          id="example-modal-1"
          aria-labelledby="modal-1-heading"
          aria-describedby="modal-1-description"
        >
          <div className="usa-modal-overlay" aria-controls="example-modal-1">
            <div
              className={`usa-modal ${
                modalType === "large" ? "usa-modal--lg" : ""
              }`}
              id="example-modal-1"
              aria-labelledby="modal-1-heading"
              aria-describedby="modal-1-description"
            >
              <div className="usa-modal__content">
                <div className="usa-modal__main">
                  <h2 className="usa-modal__heading" id="modal-1-heading">
                    {heading}
                  </h2>
                  <div className="usa-prose">
                    <p id="modal-1-description">{description}</p>
                  </div>
                  <div className="usa-modal__footer">
                    <ul className="usa-button-group">
                      <li className="usa-button-group__item">
                        <Button
                          buttonText={primaryButtonText}
                          onClick={primaryOnClick}
                        />
                      </li>
                      <li>
                        <Button
                          buttonText={secondaryButtonText}
                          buttonVariation="link"
                          type="button"
                          className="usa-button--unstyled padding-105"
                          onClick={secondaryOnClick}
                        />
                      </li>
                    </ul>
                  </div>
                </div>
                {modalType !== "forcedAction" && (
                  <button
                    type="button"
                    className="usa-button usa-modal__close"
                    aria-label="Close this window"
                    data-close-modal
                    onClick={onClose}
                  >
                    <Icon name="close" />
                  </button>
                )}
              </div>
            </div>
          </div>
        </div>
      )}
    </>
  );
};
