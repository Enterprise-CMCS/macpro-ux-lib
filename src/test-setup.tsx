import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
