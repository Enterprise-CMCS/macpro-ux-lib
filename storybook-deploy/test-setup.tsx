import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
