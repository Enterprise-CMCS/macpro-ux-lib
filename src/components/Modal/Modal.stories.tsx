import React, { useState } from "react";
import { Button } from "../Button/Button";
import { ComponentMeta, ComponentStory } from "@storybook/react";
import { Modal } from "./Modal";

export default {
  title: "Components/Modal",
  component: Modal,
  argTypes: {
    modalType: {
      description:
        "The type of Modal. Options include `default`, `large`, or `forcedAction`.",
    },
  },
  args: {
    description: "You have unsaved changes that will be lost.",
    heading: "Are you sure you want to continue?",
    modalType: "default",
    primaryButtonText: "Primary Button",
    secondaryButtonText: "Secondary Button",
  },
} as ComponentMeta<typeof Modal>;

const Template: ComponentStory<typeof Modal> = ({ ...args }) => {
  const [isOpen, setOpen] = useState(false);
  const toggleModal = () => {
    setOpen(!isOpen);
  };

  return (
    <>
      <Button buttonText="Open deftaul modal" onClick={toggleModal} />
      <Modal
        {...args}
        isOpen={isOpen}
        onClose={toggleModal}
        primaryOnClick={toggleModal}
        secondaryOnClick={toggleModal}
      />
    </>
  );
};

export const Default = Template.bind({});
Default.args = {};
