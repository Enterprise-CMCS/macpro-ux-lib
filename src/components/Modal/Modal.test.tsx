import { Modal } from "./Modal";
import { fireEvent, render, screen } from "../../test-setup";
import React from "react";

describe("Modal", () => {
  let mockAction = jest.fn();

  describe("Default", () => {
    beforeEach(() => {
      mockAction.mockClear();

      render(
        <Modal
          description="You have unsaved changes that will be lost."
          heading="Are you sure you want to continue?"
          id="example-modal"
          isOpen={true}
          onClose={mockAction}
          primaryButtonText="Continue without saving"
          primaryOnClick={mockAction}
          secondaryButtonText="Go back"
          secondaryOnClick={mockAction}
        />
      );
    });

    it("renders the heading text", () => {
      const m = screen.getByText("Are you sure you want to continue?");
      expect(m).toBeInTheDocument();
    });

    it("renders the description text", () => {
      const m = screen.getByText("You have unsaved changes that will be lost.");
      expect(m).toBeInTheDocument();
    });

    it("renders the button text", () => {
      const primary = screen.getByText("Continue without saving");
      expect(primary).toBeInTheDocument();

      const secondary = screen.getByText("Go back");
      expect(secondary).toBeInTheDocument();
    });

    it("renders the close button", () => {
      const close = screen.getByLabelText("Close this modal");
      expect(close).toBeInTheDocument();
    });

    it("calls onClose() callback", () => {
      const close = screen.getByLabelText("Close this modal");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(close);
      expect(mockAction).toBeCalled();
    });

    it("calls primaryOnClick() callback", () => {
      const primary = screen.getByText("Continue without saving");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(primary);
      expect(mockAction).toBeCalled();
    });

    it("calls secondaryOnClick() callback", () => {
      const secondary = screen.getByText("Go back");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(secondary);
      expect(mockAction).toBeCalled();
    });
  });

  describe("Large", () => {
    beforeEach(() => {
      mockAction.mockClear();

      render(
        <Modal
          description="You have unsaved changes that will be lost."
          heading="Are you sure you want to continue?"
          id="large-modal"
          isOpen={true}
          modalType="large"
          onClose={mockAction}
          primaryButtonText="Continue without saving"
          primaryOnClick={mockAction}
          secondaryButtonText="Go back"
          secondaryOnClick={mockAction}
        />
      );
    });

    it("renders the heading text", () => {
      const m = screen.getByText("Are you sure you want to continue?");
      expect(m).toBeInTheDocument();
    });

    it("renders the description text", () => {
      const m = screen.getByText("You have unsaved changes that will be lost.");
      expect(m).toBeInTheDocument();
    });

    it("renders the button text", () => {
      const primary = screen.getByText("Continue without saving");
      expect(primary).toBeInTheDocument();

      const secondary = screen.getByText("Go back");
      expect(secondary).toBeInTheDocument();
    });

    it("renders the close button", () => {
      const close = screen.getByLabelText("Close this modal");
      expect(close).toBeInTheDocument();
    });

    it("calls onClose() callback", () => {
      const close = screen.getByLabelText("Close this modal");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(close);
      expect(mockAction).toBeCalled();
    });

    it("calls primaryOnClick() callback", () => {
      const primary = screen.getByText("Continue without saving");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(primary);
      expect(mockAction).toBeCalled();
    });

    it("calls secondaryOnClick() callback", () => {
      const secondary = screen.getByText("Go back");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(secondary);
      expect(mockAction).toBeCalled();
    });
  });

  describe("Forced action", () => {
    beforeEach(() => {
      mockAction.mockClear();

      render(
        <Modal
          description="You have unsaved changes that will be lost."
          heading="Are you sure you want to continue?"
          id="forced-action-modal"
          isOpen={true}
          modalType="forcedAction"
          primaryButtonText="Continue without saving"
          primaryOnClick={mockAction}
          secondaryButtonText="Go back"
          secondaryOnClick={mockAction}
        />
      );
    });

    it("renders the heading text", () => {
      const m = screen.getByText("Are you sure you want to continue?");
      expect(m).toBeInTheDocument();
    });

    it("renders the description text", () => {
      const m = screen.getByText("You have unsaved changes that will be lost.");
      expect(m).toBeInTheDocument();
    });

    it("renders the button text", () => {
      const primary = screen.getByText("Continue without saving");
      expect(primary).toBeInTheDocument();

      const secondary = screen.getByText("Go back");
      expect(secondary).toBeInTheDocument();
    });

    it("renders the close button", () => {
      const close = screen.queryByLabelText("Close this modal");
      expect(close).toBeNull();
    });

    it("calls primaryOnClick() callback", () => {
      const primary = screen.getByText("Continue without saving");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(primary);
      expect(mockAction).toBeCalled();
    });

    it("calls secondaryOnClick() callback", () => {
      const secondary = screen.getByText("Go back");
      expect(mockAction).not.toBeCalled();
      fireEvent.click(secondary);
      expect(mockAction).toBeCalled();
    });
  });

  describe("Closed by default", () => {
    beforeEach(() => {
      render(
        <Modal
          description="You have unsaved changes that will be lost."
          heading="Are you sure you want to continue?"
          id="closed-modal"
          primaryButtonText="Continue without saving"
          primaryOnClick={mockAction}
          secondaryButtonText="Go back"
          secondaryOnClick={mockAction}
        />
      );
    });
    it("renders the heading text", () => {
      const m = screen.queryByText("Are you sure you want to continue?");
      expect(m).toBeNull();
    });

    it("renders the description text", () => {
      const m = screen.queryByText(
        "You have unsaved changes that will be lost."
      );
      expect(m).toBeNull();
    });

    it("renders the button text", () => {
      const primary = screen.queryByText("Continue without saving");
      expect(primary).toBeNull();

      const secondary = screen.queryByText("Go back");
      expect(secondary).toBeNull();
    });

    it("renders the close button", () => {
      const close = screen.queryByLabelText("Close this modal");
      expect(close).toBeNull();
    });
  });

  describe("Snapshot tests", () => {
    it("renders default Modal as expected", () => {
      const { container } = render(
        <Modal
          description="You have unsaved changes that will be lost."
          heading="Are you sure you want to continue?"
          id="example-modal"
          isOpen={true}
          onClose={mockAction}
          primaryButtonText="Continue without saving"
          primaryOnClick={mockAction}
          secondaryButtonText="Go back"
          secondaryOnClick={mockAction}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders large Modal as expected", () => {
      const { container } = render(
        <Modal
          description="You have unsaved changes that will be lost."
          heading="Are you sure you want to continue?"
          id="large-modal"
          isOpen={true}
          modalType="large"
          onClose={mockAction}
          primaryButtonText="Continue without saving"
          primaryOnClick={mockAction}
          secondaryButtonText="Go back"
          secondaryOnClick={mockAction}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("renders forced-action Modal as expected", () => {
      const { container } = render(
        <Modal
          description="You have unsaved changes that will be lost."
          heading="Are you sure you want to continue?"
          id="forced-action-modal"
          isOpen={true}
          modalType="forcedAction"
          primaryButtonText="Continue without saving"
          primaryOnClick={mockAction}
          secondaryButtonText="Go back"
          secondaryOnClick={mockAction}
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
