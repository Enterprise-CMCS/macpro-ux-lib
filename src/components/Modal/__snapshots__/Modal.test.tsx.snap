// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Snapshot tests renders default Modal as expected 1`] = `
<div>
  <div
    aria-describedby="example-modal-description"
    aria-labelledby="example-modal-heading"
    class="usa-modal-wrapper"
    id="example-modal"
    role="dialog"
  >
    <div
      aria-controls="example-modal"
      class="usa-modal-overlay"
    >
      <div
        aria-describedby="example-modal-description"
        aria-labelledby="example-modal-heading"
        class="usa-modal "
        id="example-modal"
      >
        <div
          class="usa-modal__content"
        >
          <div
            class="usa-modal__main"
          >
            <h2
              class="usa-modal__heading"
              id="modal-1-heading"
            >
              Are you sure you want to continue?
            </h2>
            <div
              class="usa-prose"
            >
              <p
                id="modal-1-description"
              >
                You have unsaved changes that will be lost.
              </p>
            </div>
            <div
              class="usa-modal__footer"
            >
              <ul
                class="usa-button-group"
              >
                <li
                  class="usa-button-group__item"
                >
                  <button
                    aria-label="Continue without saving button"
                    class="display-flex usa-button usa-button-- cms-primary-background "
                  >
                    <span
                      class="padding-x-05"
                    >
                      Continue without saving
                    </span>
                  </button>
                </li>
                <li>
                  <button
                    aria-label="Go back button"
                    class="usa-button--unstyled padding-105"
                    type="button"
                  >
                    <span
                      class="padding-x-05"
                    >
                      Go back
                    </span>
                  </button>
                </li>
              </ul>
            </div>
          </div>
          <button
            aria-label="Close this modal"
            class="usa-button usa-modal__close"
            data-close-modal="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              aria-label="close icon"
              class="usa-icon--size-3"
              color="black"
              fill="currentColor"
              role="img"
            >
              <use
                href="[object Object]#close"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal Snapshot tests renders forced-action Modal as expected 1`] = `
<div>
  <div
    aria-describedby="forced-action-modal-description"
    aria-labelledby="forced-action-modal-heading"
    class="usa-modal-wrapper"
    id="forced-action-modal"
    role="dialog"
  >
    <div
      aria-controls="forced-action-modal"
      class="usa-modal-overlay"
    >
      <div
        aria-describedby="forced-action-modal-description"
        aria-labelledby="forced-action-modal-heading"
        class="usa-modal "
        id="forced-action-modal"
      >
        <div
          class="usa-modal__content"
        >
          <div
            class="usa-modal__main"
          >
            <h2
              class="usa-modal__heading"
              id="modal-1-heading"
            >
              Are you sure you want to continue?
            </h2>
            <div
              class="usa-prose"
            >
              <p
                id="modal-1-description"
              >
                You have unsaved changes that will be lost.
              </p>
            </div>
            <div
              class="usa-modal__footer"
            >
              <ul
                class="usa-button-group"
              >
                <li
                  class="usa-button-group__item"
                >
                  <button
                    aria-label="Continue without saving button"
                    class="display-flex usa-button usa-button-- cms-primary-background "
                  >
                    <span
                      class="padding-x-05"
                    >
                      Continue without saving
                    </span>
                  </button>
                </li>
                <li>
                  <button
                    aria-label="Go back button"
                    class="usa-button--unstyled padding-105"
                    type="button"
                  >
                    <span
                      class="padding-x-05"
                    >
                      Go back
                    </span>
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal Snapshot tests renders large Modal as expected 1`] = `
<div>
  <div
    aria-describedby="large-modal-description"
    aria-labelledby="large-modal-heading"
    class="usa-modal-wrapper"
    id="large-modal"
    role="dialog"
  >
    <div
      aria-controls="large-modal"
      class="usa-modal-overlay"
    >
      <div
        aria-describedby="large-modal-description"
        aria-labelledby="large-modal-heading"
        class="usa-modal usa-modal--lg"
        id="large-modal"
      >
        <div
          class="usa-modal__content"
        >
          <div
            class="usa-modal__main"
          >
            <h2
              class="usa-modal__heading"
              id="modal-1-heading"
            >
              Are you sure you want to continue?
            </h2>
            <div
              class="usa-prose"
            >
              <p
                id="modal-1-description"
              >
                You have unsaved changes that will be lost.
              </p>
            </div>
            <div
              class="usa-modal__footer"
            >
              <ul
                class="usa-button-group"
              >
                <li
                  class="usa-button-group__item"
                >
                  <button
                    aria-label="Continue without saving button"
                    class="display-flex usa-button usa-button-- cms-primary-background "
                  >
                    <span
                      class="padding-x-05"
                    >
                      Continue without saving
                    </span>
                  </button>
                </li>
                <li>
                  <button
                    aria-label="Go back button"
                    class="usa-button--unstyled padding-105"
                    type="button"
                  >
                    <span
                      class="padding-x-05"
                    >
                      Go back
                    </span>
                  </button>
                </li>
              </ul>
            </div>
          </div>
          <button
            aria-label="Close this modal"
            class="usa-button usa-modal__close"
            data-close-modal="true"
            type="button"
          >
            <svg
              aria-hidden="true"
              aria-label="close icon"
              class="usa-icon--size-3"
              color="black"
              fill="currentColor"
              role="img"
            >
              <use
                href="[object Object]#close"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
